src/v/coproc/router.cc

```diff
@@ -112,19 +125,32 @@ ss::future<> router::do_route() {
           }
           return acc;
       };
-    return ss::map_reduce(
-             _sources.begin(),
-             _sources.end(),
-             [this](
-               std::pair<const model::ntp, ss::lw_shared_ptr<topic_state>>& p) {
-                 return route_ntp(p.first, p.second);
-             },
-             std::vector<process_batch_request::data>(),
-             std::move(reducer))
-      .then([this](std::vector<process_batch_request::data> batch) {
-          return process_batch(std::move(batch));
-      })
-      .then([this, next_loop = _jitter()] { _loop_timer.rearm(next_loop); });
+    auto mapper =
+      [this](std::pair<const model::ntp, ss::lw_shared_ptr<topic_state>>& p) {
+          return make_reader_cfg(p.second).then(
+            [this, ntp = p.first, ts = p.second](auto config) mutable {
```

Comment:
can you take `p` as a value and `p{std::move(p)}` instead of copying?

is `mapper` polymorphic (avoid `auto`) in lambda parameters.

Reply:
this was changed in a later commit, i don't want to move the shared_ptr out of the main collection though. I think this should be the only place where the topic_state shared_ptr is explicitly copied.

Reply:
sounds good. i think this is the chunk of code that is reorganized in the next patch?

Reply:
yeah exactly, and it looks like in that next patch too i still have the copy of p.